File: backend/src/trading/brokers/fyersBroker.ts
import axios from "axios";
import { BaseBroker } from "./baseBroker";
import { settings } from "../../core/config";
import { logger } from "../../core/logger";
import redisClient from "../../core/redis";

export class FyersBroker extends BaseBroker {
  private async getAuthHeaders() {
    const tokenDataRaw = await redisClient.get("fyers_access_token");
    if (!tokenDataRaw) {
      throw new Error("Fyers access token not found. Please login via /api/v1/broker/fyers/login");
    }
    const tokenData = JSON.parse(tokenDataRaw);
    const accessToken = tokenData.access_token;

    return {
      "Authorization": `${settings.FYERS_APP_ID}:${accessToken}`,
      "Content-Type": "application/json",
    };
  }

  async placeOrder(symbol: string, side: string, quantity: number, price: number): Promise<any> {
    const headers = await this.getAuthHeaders();
    const fyersSide = side.toUpperCase() === "BUY" ? 1 : -1;

    const data = {
      symbol,
      qty: Math.floor(quantity),
      type: 2, // Limit order
      side: fyersSide,
      productType: "INTRADAY",
      limitPrice: price,
      stopPrice: 0,
      validity: "DAY",
      disclosedQty: 0,
      offlineOrder: false,
    };

    try {
      const response = await axios.post("https://api-t1.fyers.in/api/v3/orders", data, { headers });
      if (response.data.s !== "ok") {
        logger.error(`Fyers order failed: ${JSON.stringify(response.data)}`);
        throw new Error(`Fyers Order Failed: ${response.data.message}`);
      }

      const orderId = response.data.id;
      const order = {
        order_id: orderId,
        symbol,
        side,
        quantity,
        price,
        status: "SUBMITTED",
      };

      logger.info(`Fyers Order Placed: ${JSON.stringify(order)}`);
      return order;
    } catch (err: any) {
      logger.error(`Fyers Place Order exception: ${err.message}`);
      throw err;
    }
  }

  async cancelOrder(orderId: string): Promise<void> {
    const headers = await this.getAuthHeaders();
    try {
      const response = await axios.delete("https://api-t1.fyers.in/api/v3/orders", {
        headers,
        data: { id: orderId },
      });
      if (response.data.s !== "ok") {
        logger.error(`Fyers cancel failed: ${JSON.stringify(response.data)}`);
        throw new Error(`Fyers Cancel Failed: ${response.data.message}`);
      }
      logger.info(`Fyers Order Cancelled: ${orderId}`);
    } catch (err: any) {
      logger.error(`Fyers Cancel Order exception: ${err.message}`);
      throw err;
    }
  }

  async getPositions(): Promise<any[]> {
    const headers = await this.getAuthHeaders();
    try {
      const response = await axios.get("https://api-t1.fyers.in/api/v3/positions", { headers });
      if (response.data.s !== "ok") {
        throw new Error(`Failed to fetch Fyers positions: ${response.data.message}`);
      }

      const fyersPositions = response.data.netPositions || [];
      return fyersPositions.map((p: any) => {
        const qty = p.netQty || 0;
        let side = "CLOSED";
        if (qty > 0) side = "BUY";
        else if (qty < 0) side = "SELL";

        return {
          symbol: p.symbol,
          quantity: Math.abs(qty),
          side,
          average_price: p.avgPrice,
        };
      });
    } catch (err: any) {
      logger.error(`Fyers getPositions exception: ${err.message}`);
      throw err;
    }
  }

  async getBalance(): Promise<{ balance: number }> {
    const headers = await this.getAuthHeaders();
    try {
      const response = await axios.get("https://api-t1.fyers.in/api/v3/funds", { headers });
      if (response.data.s !== "ok") {
        throw new Error(`Failed to fetch Fyers funds: ${response.data.message}`);
      }

      const fundLimits = response.data.fund_limit || [];
      let availableBalance = 0.0;
      for (const fund of fundLimits) {
        if (fund.title === "Available Balance") {
          availableBalance = fund.equityAmount || 0.0;
          break;
        }
      }

      return {
        balance: availableBalance,
      };
    } catch (err: any) {
      logger.error(`Fyers getBalance exception: ${err.message}`);
      throw err;
    }
  }
}

BaseBroker.register("fyers", FyersBroker);
export default FyersBroker;

File: legacy/python-backend/app/brokers/fyers_broker.py
import uuid
from fyers_apiv3 import fyersModel

from app.brokers.interfaces.base_broker import BaseBroker
from app.services.broker.fyers_token_service import FyersTokenService
from app.core.config import settings
from app.core.logger import logger


class FyersBroker(BaseBroker):
    
    broker_name = "fyers"

    async def _get_fyers_client(self):
        token_data = await FyersTokenService.get_token()
        if not token_data or "access_token" not in token_data:
            raise ValueError("Fyers access token not found. Please login via /api/v1/broker/fyers/login")
        
        access_token = token_data["access_token"]
        
        fyers = fyersModel.FyersModel(
            client_id=settings.FYERS_APP_ID,
            is_async=False,
            token=access_token,
            log_path=""
        )
        return fyers

    async def place_order(self, symbol, side, quantity, price):
        fyers = await self._get_fyers_client()
        
        # Fyers requires a specific side format: 1 for BUY, -1 for SELL
        fyers_side = 1 if side.upper() == "BUY" else -1
        
        data = {
            "symbol": symbol,
            "qty": int(quantity),
            "type": 2, # 2 is for Limit Order, 1 for Market
            "side": fyers_side,
            "productType": "INTRADAY",
            "limitPrice": float(price),
            "stopPrice": 0,
            "validity": "DAY",
            "disclosedQty": 0,
            "offlineOrder": False,
        }
        
        response = fyers.place_order(data=data)
        
        if response.get("s") != "ok":
            logger.error(f"Fyers order failed: {response}")
            raise Exception(f"Fyers Order Failed: {response.get('message')}")
            
        order_id = response.get("id")
        
        order = {
            "order_id": order_id,
            "symbol": symbol,
            "side": side,
            "quantity": quantity,
            "price": price,
            "status": "SUBMITTED" # Real status would come via websocket or polling
        }
        
        logger.info(f"Fyers Order Placed: {order}")
        return order

    async def cancel_order(self, order_id):
        fyers = await self._get_fyers_client()
        response = fyers.cancel_order(data={"id": order_id})
        
        if response.get("s") != "ok":
            logger.error(f"Fyers cancel failed: {response}")
            raise Exception(f"Fyers Cancel Failed: {response.get('message')}")
            
        logger.info(f"Fyers Order Cancelled: {order_id}")

    async def get_positions(self):
        fyers = await self._get_fyers_client()
        response = fyers.positions()
        
        if response.get("s") != "ok":
            raise Exception(f"Failed to fetch Fyers positions: {response.get('message')}")
            
        fyers_positions = response.get("netPositions", [])
        positions = []
        for p in fyers_positions:
            qty = p.get("netQty", 0)
            side = "CLOSED"
            if qty > 0:
                side = "BUY"
            elif qty < 0:
                side = "SELL"
                
            positions.append({
                "symbol": p.get("symbol"),
                "quantity": abs(qty),
                "side": side,
                "average_price": p.get("avgPrice")
            })
            
        return positions

    async def get_balance(self):
        fyers = await self._get_fyers_client()
        response = fyers.funds()
        
        if response.get("s") != "ok":
            raise Exception(f"Failed to fetch Fyers funds: {response.get('message')}")
            
        fund_limits = response.get("fund_limit", [])
        available_balance = 0.0
        for fund in fund_limits:
            if fund.get("title") == "Available Balance":
                available_balance = fund.get("equityAmount", 0.0)
                break
                
        return {
            "balance": available_balance
        }
